workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/DialogCollator.kt

```diff
@@ -0,0 +1,266 @@
+   * [DialogSession] to show it.
+   *
+   * @param onSessionsUpdated called immediately after the given [updates] are applied.
+   * Provides the updated list of [DialogSession] that should replace those that
+   * were provided to [establishDialogCollator].
+   */
+  internal fun scheduleUpdates(
+    id: UUID,
+    updates: List<DialogSessionUpdate>,
+    onSessionsUpdated: (List<DialogSession>) -> Unit
+  ) {
+    check(expectedUpdates > 0) {
+      "Each update() call must be preceded by a call to ViewEnvironment.establishDialogCollator, " +
+        "but expectedUpdates is $expectedUpdates"
+    }
+
+    this.allUpdates.add(IdAndUpdates(id, updates, onSessionsUpdated))
+    if (--expectedUpdates == 0) doUpdate()
+  }
+
+  private fun getUpdatedTopModalIndex(): Int {
+    return allUpdates.asSequence()
+      .flatMap { it.updates.asSequence().map { update -> update.overlay } }
+      .indexOfLast { it is ModalOverlay }
+  }
+
+  private fun doUpdate() {
+    // Flatten establishedSessions into an Iterator across the existing sessions from
+    // bottom to top, each paired with the id of the LayeredDialogSessions that registered it.
+    val establishedSessionsIterator: Iterator<Pair<UUID, DialogSession>> =
+      establishedSessions.asReversed()
+        .asSequence()
+        .flatMap { it.sessions.map { session -> Pair(it.id, session) } }
+        .iterator()
+
+    // Collects members of establishedSessions that are dismissed because they
+    // are out of order, so that we can try to show them again.
+    val dismissedSessions = mutableListOf<Pair<UUID, DialogSession>>()
+
+    // Z index of the uppermost ModalOverlay.
+    val topModalIndex = getUpdatedTopModalIndex()
+
+    // Z index of the dialog session being updated.
+    var updatingSessionIndex = 0
+
+    allUpdates.forEach { idAndUpdates ->
+      val updatedSessions = mutableListOf<DialogSession>()
+
+      // We're building an object that the next LayeredDialogSessions can use to
+      // find an existing dialog that matches a given Overlay. Any
+      // incompatible dialog that we skip on the way to find a match is dismissed.
+      // If we later find a match for one of the dismissed dialogs, it is re-shown --
+      // which moves it to the front, ensuring the correct Z order.
+      val oldSessionProvider = OldSessionProvider { overlay ->
+
+        // First we iterate through the existing windows to find one that belongs
+        // to this group and matches the overlay.
+        while (establishedSessionsIterator.hasNext()) {
+          val (id, session) = establishedSessionsIterator.next()
+          if (idAndUpdates.id == id && session.canShow(overlay)) return@OldSessionProvider session
+
+          // Can't update this session from this Overlay. Dismiss it (via Dialog.dismiss
+          // under the hood), but hold on to it in case it can except a later Overlay.
+          session.setVisible(false)
```

Comment:
here and in the val we call them `dismissed` - but that's an implementation detail we're trying to hide with `setVisible`.

Would it be more helpful to call them `hiddenSessions`?

Reply:
Heh. It was `hiddenSessions` originally but I worried that was too opaque. Going back to it.